lib/diagnostics_channel.js

```diff
@@ -0,0 +1,281 @@
+'use strict';
+
+const {
+  ArrayPrototypeIndexOf,
+  ArrayPrototypePush,
+  ArrayPrototypeShift,
+  ArrayPrototypeSplice,
+  ObjectCreate,
+  ObjectSetPrototypeOf,
+  Map,
+  MapPrototypeDelete,
+  MapPrototypeGet,
+  MapPrototypeSet,
+  Promise,
+  PromiseResolve,
+  SymbolAsyncIterator
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+  }
+} = require('internal/errors');
+
+const { triggerUncaughtException } = internalBinding('errors');
+
+const { WeakReference } = internalBinding('util');
+
+function identity(v) {
+  return v;
+}
+
+class AsyncIterableChannel {
+  constructor(channel) {
+    this.channel = channel;
+    this.events = [];
+    this.waiting = [];
+
+    this.subscriber = (message) => {
+      const resolve = ArrayPrototypeShift(this.waiting);
+      if (resolve) {
+        return resolve({
+          value: message,
+          done: false
+        });
+      }
+
+      ArrayPrototypePush(this.events, message);
+    };
+
+    channel.subscribe(this.subscriber);
+  }
+
+  [SymbolAsyncIterator]() {
+    return this;
+  }
+
+  return() {
+    const data = { done: true };
+    this.done = true;
+
+    this.channel.unsubscribe(this.subscriber);
+
+    for (let i = 0; i < this.waiting.length; i++) {
+      const resolve = this.waiting[i];
+      resolve(data);
+    }
+
+    return PromiseResolve(data);
+  }
+
+  next() {
+    const event = ArrayPrototypeShift(this.events);
+    if (event) {
+      return PromiseResolve({
+        value: event,
+        done: false
+      });
+    }
+
+    if (this.done) {
+      return PromiseResolve({
+        done: true
+      });
+    }
+
+    return new Promise((resolve) => {
+      ArrayPrototypePush(this.waiting, resolve);
+    });
+  }
+}
+
+class SpanMessage {
+  constructor(action, id, data) {
+    this.action = action;
+    this.id = id;
+    this.data = data;
+  }
+
+  // Augment with type: span to identify over the wire through JSON
+  toJSON() {
```

Comment:
The name of this function is misleading as the data could contain e.g. a `Buffer` which doesn't serialize well to JSON.

Reply:
That's intentional, so it gets called if you do `JSON.stringify(spanMessage)`. I agree it's _possible_ it could contain a buffer, so I'd encourage publishers to avoid complex types where possible.

Reply:
I don't think serialization of data should be part of this module. I expect a lot users will serialize only a part of the data published.
For example in your http server sample you publish req/res objects which is not what someone would send out as is.

Reply:
In many cases they would. A specific agent design I intend for this to enable is one where span lifecycle events are sent out of the process _immediately_ and the aggregate happens in another process, significantly reducing the overhead of a production monitoring agent. Regardless of if I include a `toJSON`, users _will_ try to send this data over the wire. All this function does it make it a little easier to keep the identity over the wire. It's not intended to imply that it is _safe_ to just pipe everything over the wire as-is, and is therefore not currently documented, but I'd rather keep this just for the safety of properly identifying span messages from over the wire than to omit it and have users complaining that the `aggregate` function isn't finding their spans. 🤷

Reply:
I just tries to use `JSON.stringify` for the data provided in you http_server sample. It results in `Uncaught TypeError: Converting circular structure to JSON`.

If diagnostics channel includes support for serializing to JSON we should clearly document which data is allowed to be published.

I'm not sure if serializing and sending all data is really reducing overhead. In general I like you approach in http_server that you publish existing, well documented objects via dc. This avoids the overhead to extract and pre-format the data and reduces the risk hat user A misses foo and user B misses bar. Different tools will have different needs on the data and as publisher I would tend to publish more data as filtering is simpler then adding something.

Reply:
As I said, it's specifically _not_ documented because I don't want to encourage users to do this, but I know they _will_ so I'd prefer to be slightly less likely for them to shoot themselves in the foot.

Reply:
I've pulled all the span stuff out to a separate commit, and we may consider dropping it entirely or making it a follow-up PR, if we can't reach consensus here on that part.

Reply:
Marking as resolved as I've removed the span stuff from this PR. I will open a new PR to add that stuff separately, so feel free to bring this up again there.